### PR TITLE
all-packages.nix: cleanup, remove `res.` (next stage super)

### PR DESCRIPTION
**Pull request by oxij**
# What? Why?

#55061, but with tests still evaluating, as far as I can see.

# `git log`

- libxml2Python: move from all-packages.nix

- all-packages.nix: kill all noop uses of `res`

# `nix-instantiate` environment

- Host OS: Linux 4.9, SLNOS 19.03
- Nix: nix-env (Nix) 2.1.3
- Multi-user: yes
- Sandbox: yes
- NIXPKGS_CONFIG:

```nix
{
  checkMeta = true;
  doCheckByDefault = true;
}
```

# `nix-env -qaP` diffs

- On x86_64-linux: noop
- On aarch64-linux: noop
- On x86_64-darwin: noop

/cc @Ericson2314 @danbst @Infinisil from #55061

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -4823,27 +4823,29 @@ in
 
   phodav = callPackage ../tools/networking/phodav { };
 
-  pinentry = callPackage ../tools/security/pinentry {
+  pinentry-real = callPackage ../tools/security/pinentry {
```
Comment (matthewbauer) [inline pkgs/top-level/all-packages.nix]: Any reason for doing this? I'd prefer to not introduce any new names like this.

Comment (oxij) [reply pkgs/top-level/all-packages.nix]: See https://github.com/NixOS/nixpkgs/pull/55061#issuecomment-460108869 and above discussion.

Comment (matthewbauer) [reply pkgs/top-level/all-packages.nix]: I guess I'd rather have res. than have to deal with the -real derivations. You could also fix the no-x-libs.nix thing by just overriding like:

```
pinentry = super.pinentry.override {
  gtk2 = null;
}
```

in no-x-libs.nix

**Comment by danbst** [comment]
The problem is not with just removing `res`, but with overrides in general. I've described an example at https://github.com/NixOS/nixpkgs/pull/55061#issuecomment-460052013, which this PR doesn't touch at all (but problem exists!)

I imagine another solution (at least for pinentry case):
```
inherit (rec {
  pinentry = callPackage ../tools/security/pinentry {
    libcap = if stdenv.isDarwin then null else libcap;
  };
  pinentry_ncurses = pinentry.override {
    gtk2 = null;
  };

  pinentry_emacs = pinentry.override {
    enableEmacs = true;
  };

  pinentry_gnome = pinentry.override {
    inherit gcr;
  };

  pinentry_qt4 = pinentry.override {
    qt = qt4;
  };

  pinentry_qt5 = pinentry.override {
    qt = qt5.qtbase;
  };  
})
  pinentry
  pinentry_ncurses
  pinentry_emacs
  pinentry_gnome
  pinentry_qt4
  pinentry_qt5
;
```
In that case there will be no common ancestor, so overrides will behave better (for a while). It is also slightly more verbose, to discourage such pattern use in `all-packages.nix`.

The changes, which don't use `override` are fine IMO.

**Comment by Ericson2314** [comment]
@matthewbauer I don't like `res` because it's such a global solution to a local problem: nothing else that doesn't need it should be confused by some big extra fix point.

I don't think `super` will work because the tests would add an overlay after, and pinentry is defined in the same overlay as the super reference so the super will be resolved by "looping back around" to hit the final self. (I'm not totally sure I got this right though.)

The `*-real` style solution is nice because while it does leak a new name, this allow overriding all at once, which is nice if one wants do a non-variant-specific change. I suggested the "real" suffix because I thought of `pinentry` as a metapackage like `libc` vs `glibc`. But since there's only one pinentry, but many configuration options, something like `pinentryBase` might be better.

If we don't care about overriding all at once, then @danbst's is the best solution of all as it's truely local. Great idea @danbst!!

**Comment by matthewbauer** [comment]
> I don't like `res` because it's such a global solution to a local problem: nothing else that doesn't need it should be confused by some big extra fix point.

It doesn't sound that local to me. It sounds like anything using an override (or even aliases) within all-packages.nix could run into this problem. @danbst would work okay, but this would be a lot of recs introduced to support all of the override cases.

My ideal solution would be to move all of these overrides to ```pkgs/top-level/overrides.nix``` and make that a proper overlay. I had a PR to do this but I can't find it right now (and it's probably out of date). I think there were some objections to that though.

Otherwise, maybe we should just move back to using callPackage multiple times? I prefer using override because it makes it more clear what the "true" package is but we don't want to have this kind of thing broken...

**Comment by Ericson2314** [comment]
Err my local I meant each instance of the problem, i.e. only the other pinentries need to worry about the base one, be it be given its own name or thrown away as a temporary.

**Comment by oxij** [comment]
@danbst's hack is cute, but if not the `-real` suffix then I vote for pure `callPackage` solution. Anything that is duplicated just needs to be moved to the package expression itself anyway.

**Comment by danbst** [comment]
agree about callPackage and moving common stuff to package itself.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (matthewbauer) [reply pkgs/top-level/all-packages.nix]: i suppose this isn't so bad for now. Could you call it pinentry_full to a be a little more descriptive?

Comment (matthewbauer) [reply pkgs/top-level/all-packages.nix]: What worries me though is having to have a "real" version for every package with an override. That looks like it is needed for the ```a = a_<flavor>;``` type overrides to work.

Comment (matthewbauer) [reply pkgs/top-level/all-packages.nix]: Another solution would be to change this line in no-x-libs.nix from

```
      pinentry = super.pinentry_ncurses;
```

to

```
pinentry = super.pinentry.override { gtk2 = null; }
```

**Review by matthewbauer — APPROVED** [review]

**Comment by Ericson2314** [comment]
@matthewbauer did you change your midn on this?

**Comment by jtojnar** [comment]
Regarding pinentry, https://github.com/NixOS/nixpkgs/pull/49270 will split it into multiple outputs instead of overrides.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (roberth) [reply pkgs/top-level/all-packages.nix:4826]: This has been replaced by a multiple outputs on a single derivation. While that's horrible for _derivation_ closure size and rebuilds, it makes this discussion irrelevant for now.

**Comment by AndersonTorres** [comment]
Closing as dead.

**Pull request closed without merging**